### PR TITLE
"for... in" should not be used with arrays. This causes bugs!

### PR DESCRIPTION
**Pull request by agamemnus**
See: https://github.com/katspaugh/wavesurfer.js/issues/1014.

I also changed "entry" to "canvas" since they are canvases. And, uh, just a few styling ideas.

### CODE REVIEW

**Comment by katspaugh** [comment]
Could you just replace the `for-in` with a `forEach` without the other stylistic changes? Btw, `forEach` takes a third argument which is the context of the callback, so you could pass `this` and not introduce another `var my = this` to keep the context.

**Comment by agamemnus** [comment]
I already ended up removing many of those changes due to your draconian style checker! The rest is as follows --

There are only four differences now, other than the forEach fixes, and corresponding "my" variables -- the "my" variables isn't something I came up with as it was present in an existing function.
1) Some brackets had an errant starting space: e.g.: `{ width: elementWidth + 'px'}`. and the `getImage` function didn't have a space between it and the parameters.
2) `entry` -> `canvas` and `lastEntry` -> `lastCanvas`. IMO, if you have `this.canvases`, then one of them should be a `canvas` and not an `entry`.
3) Some breaking up of a multi-line var:

````
var elementWidth = Math.round(width / this.params.pixelRatio),
totalWidth = Math.round(this.width / this.params.pixelRatio);
````
to
````
var elementWidth = Math.round(width / this.params.pixelRatio);
var totalWidth   = Math.round(this.width / this.params.pixelRatio);
````
and
````
var entry = {},
leftOffset = this.maxCanvasElementWidth * this.canvases.length;
````
to
````
var canvas = {};
var leftOffset = this.maxCanvasElementWidth * this.canvases.length;
````
4)
````
var endCanvas = Math.min(Math.ceil((x + width) / this.maxCanvasWidth) + 1,
this.canvases.length);
````
to
````
var endCanvas = Math.min(Math.ceil((x + width) / this.maxCanvasWidth) + 1, this.canvases.length);
````

**Comment by katspaugh** [comment]
While some of your changes make sense (I'll mark them with a thumbs-up in the changeset), some are arbitrary, and I would rather not create more maintenance work for the `next` branch.

File: src/drawer.multicanvas.js [hunk 1]
```diff
@@ -52,24 +53,24 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
             this.removeCanvas();
         }
 
-        for (var i in this.canvases) {
+        my.canvases.forEach (function (canvas, i) {
```
Comment (katspaugh) [inline src/drawer.multicanvas.js]: 👍  except for the extra space after `forEach` – it's not needed.

File: src/drawer.multicanvas.js [hunk 2]
```diff
@@ -91,51 +92,52 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
...
     },
 
     clearWave: function () {
-        for (var i in this.canvases) {
-            this.clearWaveForEntry(this.canvases[i]);
-        }
+        var my = this;
+        my.canvases.forEach (function (canvas) {
+            my.clearWaveForEntry(canvas);
+        });
```
Comment (katspaugh) [inline src/drawer.multicanvas.js]: 👍 

```
this.canvases.forEach(function (canvas) {
    this.clearWaveForEntry(canvas);
}, this);
```

File: src/drawer.multicanvas.js [hunk 3]
```diff
@@ -223,31 +223,29 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
...
-        for (var index in this.canvases) {
-            var entry = this.canvases[index];
-
-            this.setFillStyles(entry);
-
-            this.drawLineToContext(entry, entry.waveCtx, peaks, absmax, halfH, offsetY, start, end);
-            this.drawLineToContext(entry, entry.progressCtx, peaks, absmax, halfH, offsetY, start, end);
-        }
+        var my = this;
+        my.canvases.forEach (function (canvas) {
```
Comment (katspaugh) [inline src/drawer.multicanvas.js]: The same comment as above.

File: src/drawer.multicanvas.js [hunk 4]
```diff
@@ -333,16 +330,16 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
...
     /**
-     * Combine all available canvasses together.
+     * Combine all available canvases together.
      *
      * @param {String} type - an optional value of a format type. Default is image/png.
      * @param {Number} quality - an optional value between 0 and 1. Default is 0.92.
      *
      */
-    getImage: function(type, quality) {
+    getImage: function (type, quality) {
```
Comment (katspaugh) [inline src/drawer.multicanvas.js]: :thumbsup:

**Comment by agamemnus** [comment]
OK, give me a few minutes to make the appropriate changes.

**Comment by agamemnus** [comment]
Ok, I removed almost all of the other changes.

I also implemented your suggestion about inserting a `this` variable in forEach.

**Comment by katspaugh** [comment]
I still see a lot of other changes in the diff.

**Comment by agamemnus** [comment]
I fixed some errors.

The only other change now is removing space between opening/closing brackets and text. You don't really even have that consistent...

**Comment by agamemnus** [comment]
Ok, a few more I didn't notice.

**Comment by agamemnus** [comment]
I really hate those extraneous spaces, but I added them back. I also added some extra extraneous spaces in style definition calls for it to be consistent.

File: src/drawer.multicanvas.js [hunk 5]
```diff
@@ -191,23 +190,22 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
...
+                }, this);
                 return;
             } else {
                 peaks = channels[0];
             }
         }
 
         // Support arrays without negative peaks
-        var hasMinValues = [].some.call(peaks, function (val) { return val < 0; });
+        var hasMinValues = [].some.call(peaks, function (val) {return val < 0;});
```
Comment (katspaugh) [inline src/drawer.multicanvas.js]: Could you revert this change, too?

**Comment by agamemnus** [comment]
Ok.

**Comment by katspaugh** [comment]
Merged, thanks!
